File: PaymentProcessor/PaymentProcessor.ts
export default abstract class PaymentProcessor {
  processPayment(amount: number, currency: string): void {
    this.validate();
    this.performPayment(amount, currency);
    this.sendReceipt();
  }

  abstract validate(): void;
  abstract performPayment(amount: number, currency: string): void;
  sendReceipt(): void {
    console.log("Sending Receipt");
  }
}

File: labs/lab02/lab2/templatemethodlab2/src/main/java/with/templatemethod/PaymentProcessor.java
package with.templatemethod;

import without.templatemethod.Customer;

public abstract class PaymentProcessor {
    protected Customer customer;

    public PaymentProcessor(Customer customer) {
        this.customer = customer;
    }

    public final void processPayment(double amount, String currency) {
        validate();
        performPayment(amount, currency);
        sendReceipt();
    }

    protected abstract void validate();
    protected abstract void performPayment(double amount, String currency);
    protected abstract void sendReceipt();
}
